### PR TITLE
fix: login button was disabled when closing popup in 2fa screen (VIALJS-136)

### PR DESCRIPTION
**Pull request by c0deaddict**
### Issue number

https://voipgrid.atlassian.net/browse/VIALJS-136

Related: https://github.com/vialer/vjs-adapter-user-vg/pull/2

### Expected behaviour

When user enters her username/password in the plugin and then closes the popup (to look up 2FA token for ex.) then opens the popup again and fills in her token, the Login button should be enabled.

### Actual behaviour

Login button is disabled.

### Description of fix

The password (from the 1st step of the Login view) is stored in the *fg* state, which is lost when the popup is closed. Added a callback hook in `bg:user:login` that will temporarily store the password in the *bg* state when `twoFactor == true`.

Also fixed the validation messages: they were not showing anymore.
And fixed the pre-commit hooks.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -327,17 +327,17 @@ gulp.task('manifest', 'Generate a browser-specific manifest file.', (done) => {
 
 
 gulp.task('pre-commit-lint-js', () => {
-    return guppy.stream('pre-commit')
-        .pipe(filter(['*.js']))
+    return gulp.src(guppy.src('pre-commit'))
```
Comment (ghost) [inline gulpfile.js]: Why did you change this from guppy.stream to gulp.src?

File: gulpfile.js [hunk 2]
```diff
@@ -327,17 +327,17 @@ gulp.task('manifest', 'Generate a browser-specific manifest file.', (done) => {
...
         .pipe(eslint())
         .pipe(eslint.format())
         .pipe(eslint.failAfterError())
 })
 
 
 gulp.task('pre-commit-lint-sass', () => {
-    return guppy.stream('pre-commit')
-        .pipe(filter(['*.scss']))
+    return gulp.src(guppy.src('pre-commit'))
```
Comment (ghost) [inline gulpfile.js]: Same question

File: gulpfile.js [hunk 3]
```diff
@@ -347,7 +347,7 @@ gulp.task('pre-commit-lint-sass', () => {
 
 
 gulp.task('pre-commit-protect-secrets', () => {
-    return guppy.stream('pre-commit')
+    return gulp.src(guppy.src('pre-commit'))
```
Comment (ghost) [inline gulpfile.js]: And here

File: src/components/field/index.js [hunk 4]
```diff
@@ -25,17 +25,7 @@ module.exports = (app) => {
...
-                // Validation for `requiredIf` depends on the state of other
-                // fields. Therefor don't use the $dirty check on this field,
-                // but go straight for the $invalid state.
-                if ('requiredIf' in this.validation) {
-                    return this.validation.$invalid
-                }
-
-                // Invalid has 3 states: true, false and null (not changed/dirty).
-                return this.validation.$error
+                return this.validationMessage !== ''
```
Comment (ghost) [inline src/components/field/index.js]: I think this needs more work:
- The validation message should not be triggered directly after the component is mounted
- The login button stays disabled(not validated) when closing the popup and reopening

File: src/components/login/index.js [hunk 5]
```diff
@@ -7,11 +7,16 @@ module.exports = (app) => {
     */
     const Login = {
         computed: app.helpers.sharedComputed(),
+        created: function() {
```
Comment (ghost) [inline src/components/login/index.js:10]: Just use Vue-stash store: {} to mount the store property in the component, instead of copying it to data.

File: src/js/lib/skeleton.js [hunk 6]
```diff
@@ -71,38 +71,39 @@ class Skeleton extends EventEmitter {
     * @param {Boolean|String} noIpc - Flag to skip ipc emission or to do `both`.
     * @param {Boolean|String} [tabId=false] - Emit to specific tab over ipc.
     * @param {Boolean|String} [parent=false] - Emit to script's parent over ipc.
+    * @returns {Promise} - Promise that will resolve when the emit is done - if availabl
     */
     emit(event, data = {}, noIpc = false, tabId = false, parent = false) {
```
Comment (ghost) [inline src/js/lib/skeleton.js]: Make it `async emit`, so you can use await instead of .then?

File: src/js/lib/skeleton.js [hunk 7]
```diff
@@ -71,38 +71,39 @@ class Skeleton extends EventEmitter {
...
+        let promise
```
Comment (ghost) [inline src/js/lib/skeleton.js]: Maybe wrap the whole method in a Promise() to make the return type consistent?

File: src/js/lib/skeleton.js [hunk 8]
```diff
@@ -71,38 +71,39 @@ class Skeleton extends EventEmitter {
...
         if (this.env.isExtension && (!noIpc || noIpc === 'both')) {
             let payloadData = {data: data, event: event}
 
             if (tabId) {
                 if (this.verbose) this.logger.debug(`${this}emit ipc event '${event}' to tab ${tabId}`)
-                browser.tabs.sendMessage(tabId, payloadData).catch((err) => {
+                return browser.tabs.sendMessage(tabId, payloadData).catch((err) => {
```
Comment (ghost) [inline src/js/lib/skeleton.js]: try/catch (with async)

File: src/js/lib/skeleton.js [hunk 9]
```diff
@@ -71,38 +71,39 @@ class Skeleton extends EventEmitter {
...
                     if (this.verbose) this.logger.debug(`${this}${err.message}`)
                 })
-                return
             } else if (parent) {
                 if (this.verbose) this.logger.debug(`${this}emit ipc event '${event}' to parent`)
                 parent.postMessage({data: data, event: event}, '*')
-                return
+                return undefined
```
Comment (ghost) [inline src/js/lib/skeleton.js]: resolve() ?

File: src/js/lib/skeleton.js [hunk 10]
```diff
@@ -71,38 +71,39 @@ class Skeleton extends EventEmitter {
...
                 }).catch((err) => {
                     if (this.verbose) this.logger.debug(`${this}${err.message}`)
                 })
             } else {
                 if (this.verbose) this.logger.debug(`${this}emit ipc event '${event}'`)
-                let _promise = browser.runtime.sendMessage(payloadData)
-                if (_promise) {
-                    _promise.catch((err) => {
+                promise = browser.runtime.sendMessage(payloadData)
+                if (promise) {
```
Comment (ghost) [inline src/js/lib/skeleton.js]: It looks here as if browser.runtime.sendMessage doesn't always return a Promise, which it does.

**Review by ghost — CHANGES_REQUESTED** [review]
Two issues: 
1. Mixing up local state and global state(vue-stash) 
2. Validation rules are not applied correctly (vuelidate)

File: gulpfile.js (hunk 1 above)
Comment (c0deaddict) [reply gulpfile.js]: I tried various things to get it working again, and with this is worked. But looking at the documentation I think `guppy.stream` is what we want not `guppy.src`:

> Note that depending on the git-hook, you may be acting on files that differ from your working copy, such as those staged for commit (as with 'pre-commit' for example), rather than the working copy. If you need to act on the working-copy files, use guppy.src() instead.

I'll check and see if the stream works now, maybe it was just the filter that was too narrow.
